File: sdks/python/weaveport_sdk/cleanup.py
"""Sequential owned cleanup with primary and secondary causes preserved."""

from .context import SessionCleanupError


async def cleanup_all(*actions):
    errors = []
    for action in actions:
        try:
            await action()
        except BaseException as error:
            errors.append(error)
    if errors:
        raise SessionCleanupError(
            "Owned resource cleanup failed", errors=errors
        ) from errors[0]


async def execute_with_cleanup(action, cleanup):
    primary = None
    try:
        return await action()
    except BaseException as error:
        primary = error
        raise
    finally:
        try:
            await cleanup()
        except BaseException as secondary:
            if primary is None:
                raise
            failure = SessionCleanupError(
                "Execution and cleanup failed", errors=[primary, secondary]
            )
            failure.has_execution_failure = True
            raise failure from primary

File: sdks/typescript/src/runtime/cleanup.ts
import { SessionCleanupError } from '../session.js';

export async function cleanupAll(...actions: (() => Promise<unknown>)[]): Promise<void> {
    const errors: unknown[] = [];
    for (const action of actions) {
        try {
            await action();
        } catch (error) {
            errors.push(error);
        }
    }
    if (errors.length) throw new SessionCleanupError(errors, 'Owned resource cleanup failed');
}

export async function executeWithCleanup<T>(
    action: () => Promise<T>,
    cleanup: () => Promise<void>,
): Promise<T> {
    const failures: unknown[] = [];
    try {
        return await action();
    } catch (error) {
        failures.push(error);
        throw error;
    } finally {
        try {
            await cleanup();
        } catch (error) {
            if (!failures.length) throw error;
            const combined = new SessionCleanupError(
                [...failures, error],
                'Execution and cleanup failed',
            );
            combined.hasExecutionFailure = true;
            throw combined;
        }
    }
}
